tests/apollo-fetch.ts

```diff
@@ -0,0 +1,496 @@
+      const options = fetchMock.lastCall(uri)[1];
+      const body = JSON.parse(options.body);
+      assert.deepEqual(options.method, 'POST');
+      assert.deepEqual(options.headers, {Accept: '*/*', 'Content-Type': 'application/json'});
+      assert.deepEqual(body.query, print(sampleQuery));
+      done();
+    });
+  });
+
+  it('should pass an error onto the Promise', () => {
+    const uri = 'error';
+    const fetcher = createApolloFetch({uri, customFetch: fetch});
+    const result = fetcher({query: print(sampleQuery)});
+    return assert.isRejected(result, mockError.throws, mockError.throws.message);
+  });
+
+// missingUrl
+
+  it('should catch on a network error', (done) => {
+    const fetcher = createApolloFetch({uri: unauthorizedUrl});
+    const result = fetcher({query: print(sampleQuery)});
+    result.then(expect.fail)
+    .catch((error) => {
+      assert.deepEqual(error.message, 'Network request failed with status 403 - \"Forbidden\"');
+      assert.isDefined(error.response);
+      assert.isDefined(error.parseError);
+      done();
+    });
+  });
+
+  it('should return a fail to parse response when fetch returns raw response', (done) => {
+    const fetcher = createApolloFetch({uri: '/raw'});
+    const result = fetcher({query: print(sampleQuery)});
+    result.then(expect.fail)
+    .catch((error) => {
+      assert.deepEqual(error.message, 'Network request failed to return valid JSON');
+      assert.isDefined(error.response);
+      assert.isDefined(error.raw);
+      assert.deepEqual(error.raw, unparsableData);
+      done();
+    });
+  });
+
+
+  describe('middleware', () => {
+    it('should throw an error if you pass something bad', () => {
+      const malWare: any = {};
+      const networkInterface = createApolloFetch({ uri: '/graphql' });
+
+      try {
+        networkInterface.use([malWare]);
+        expect.fail();
+      } catch (error) {
+        assert.equal(
+          error.message,
+          'Middleware must implement the applyMiddleware function',
+        );
+      }
+
+    });
+
+    it('should take a middleware and assign it', () => {
+      const testWare = TestWare();
+
+      const networkInterface = createApolloFetch({ uri: '/graphql' });
+      networkInterface.use([testWare]);
+
+      assert.equal((<any>networkInterface)._middlewares[0], testWare);
+    });
+
+    it('should take more than one middleware and assign it', () => {
+      const testWare1 = TestWare();
+      const testWare2 = TestWare();
+
+      const networkInterface = createApolloFetch({ uri: '/graphql' });
+      networkInterface.use([testWare1, testWare2]);
+
+      assert.deepEqual((<any>networkInterface)._middlewares, [testWare1, testWare2]);
+    });
+
+    it('should alter the request variables', () => {
+      const testWare1 = TestWare([
+        { key: 'personNum', val: 1 },
+      ]);
+
+      const swapi = createApolloFetch({ uri: swapiUrl });
+      swapi.use([testWare1]);
```

Comment:
It's a bit hard to know here what `testWare` does, because it's not a descriptive name.